### PR TITLE
fix Issue 23558 - add __traits(getModuleClasses [, module name])

### PR DESCRIPTION
**Pull request by WalterBright**
Provide alternative to Object.factory(), see https://github.com/dlang/dmd/pull/14681

### CODE REVIEW

**Comment by rikkimax** [comment]
I have to say, I'm not convinced by this approach for a couple of reasons. A filter approach that is more generic would be far better. ``__traits(getMembers, symbolOrModule, function/class/struct/union/field)`` would be a lot more useful and allow simplifying significant existing logic.

**Comment by WalterBright** [comment]
See also: https://www.digitalmars.com/d/archives/digitalmars/D/ModuleInfo_Object.localClasses_and_Object.find_-_any_users_366643.html

**Comment by WalterBright** [comment]
@rikkimax if you or anyone else want to write this function, please do so!

**Comment by rikkimax** [comment]
Traits at least are something I can do, so yeah it can go on the back burner after the refactor in dub to output a list of modules. Just a shame that we are introducing an ultra-specific trait when we know a more generic one will come eventually ;)

**Comment by WalterBright** [comment]
There's not much point in doing a spec PR until this is approved.

**Comment by rikkimax** [comment]
> @rikkimax if you or anyone else want to write this function, please do so!

https://issues.dlang.org/show_bug.cgi?id=23559

**Comment by ibuclaw** [comment]
But this trait won't work in the same way as the current runtime moduleinfo equivalent, will it?

```
pragma(msg, __traits(getModuleClasses));  // A? 
static if (cond) class B {} 
pragma(msg, __traits(getModuleClasses)); // A, B? 
class A { } 
enum cond = true;
```

**Comment by WalterBright** [comment]
It produces:
```
tuple("B", "test2.A")
tuple("B", "test2.A")
```

**Comment by WalterBright** [comment]
This works as is, but will have to be changed when https://github.com/dlang/dmd/pull/14708 is merged.

File: compiler/src/dmd/traits.d [hunk 1]
```diff
@@ -779,6 +781,56 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
+            {
+                if (auto imp = s.isImport())
+                    m = imp.mod;
+                else
+                    m = s.isModule();
+            }
+            if (!m)
+            {
+                e.error("in expression `%s` `%s` must be a module", e.toChars(), o.toChars());
+                return ErrorExp.get();
```
Comment (RazvanN7) [inline compiler/src/dmd/traits.d:810]: This error is not covered.

File: compiler/src/dmd/traits.d [hunk 2]
```diff
@@ -779,6 +781,56 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
+            }
+        }
+        else
+            return dimError(0);
```
Comment (RazvanN7) [inline compiler/src/dmd/traits.d:814]: Neither does this one.

**Review by RazvanN7 — APPROVED** [review]

File: compiler/src/dmd/traits.d (hunk 1 above)
Comment (WalterBright) [reply compiler/src/dmd/traits.d:810]: done

File: compiler/src/dmd/traits.d (hunk 2 above)
Comment (WalterBright) [reply compiler/src/dmd/traits.d:814]: done

**Comment by RazvanN7** [comment]
@WalterBright could you maybe add a changelog entry + a spec PR?

**Review by RazvanN7 — DISMISSED** [review]
Needs changelog entry.

**Comment by WalterBright** [comment]
@RazvanN7 ready to go

**Review by RazvanN7 — APPROVED** [review]

**Comment by adamdruppe** [comment]
> @rikkimax if you or anyone else want to write this function, please do so!

This has been done by several D programmers. It is an utterly trivial loop over the existing traits(allMembers).

**Comment by schveiguy** [comment]
This gives you the actual types (which is more useful). I wrote it in 5 minutes.
```d
template AllClasses(alias mod) {
   import std.meta : AliasSeq;
   alias result = AliasSeq!();
   static foreach(m; __traits(allMembers, mod))
      static if(is(__traits(getMember, mod, m) == class)) result = AliasSeq!(result, __traits(getMember, mod, m));
   alias AllClasses = result;
}```

**Comment by WalterBright** [comment]
Thank you, @schveiguy 

The idea is to replace the `Object.factory` and `Typeinfo_Class.find` functions that rely on `ModuleInfo`. If you want to do a better PR with your code, please do.

**Comment by schveiguy** [comment]
My point was just that the trait you created is not necessary, and in fact, not very helpful. It's hard to turn a string into a type, even when you know the module it comes from (unless you want to use `__traits(child)`, but you can't do that with this trait). The reason Object.factory works is because the *compiler* registers the types -- the strings are just used for lookup, and arbitrarily whatever the typeinfo decides to use, the important part that is needed is the storage of all the types. How do you plan to solve that?

What is needed is a complete solution of how to create a registration system. Ideally at compile-time, but if not, at least at runtime (keeping in mind that static constructors cannot be used as they add cycles to complex code).

**Comment by adamdruppe** [comment]
Let's revert this now before the useless baggage actually makes it into a release.

**Pull request merged**